boot/bootstate20.go

```diff
@@ -587,3 +690,65 @@ func (bsmark *bootState20MarkSuccessful) commit() error {
 
 	return nil
 }
+
+// genericEarlyBootChooseSnap will run the logic to choose which snap should be
+// mounted during the early boot, i.e. initramfs. Specifically, it only works
+// with kernel and base snaps. It returns the first and second choice for what
+// snaps to mount, if second is set, then it is the primary fallback snap, and
+// the first snap is the try snap, if second is unset, then first is the primary
+// fallback snap. It returns both so that a higher level function can do
+// additional verification of the try snap if it wants, such as the kernel snaps
+// being verified in the modeenv, or the base snaps updating the modeenv with
+// new values for base_status, etc.
+func genericEarlyBootChooseSnap(bs bootState, expectedTryStatus, typeString string) (
+	firstChoice, secondChoice snap.PlaceInfo,
+	err error,
+) {
+	curSnap, trySnap, snapTryStatus, err := bs.revisions()
+
+	if err != nil && !IsTrySnapError(err) {
+		// we have no fallback snap!
+		return nil, nil, fmt.Errorf("fallback %s snap unusable: %v", typeString, err)
+	}
+
+	// check that the current snap actually exists
+	file := curSnap.Filename()
+	snapPath := filepath.Join(dirs.SnapBlobDirUnder(InitramfsWritableDir), file)
+	if !osutil.FileExists(snapPath) {
+		// somehow the kernel snap on ubuntu-boot doesn't exist in ubuntu-data
```

Comment:
kernel can also be base here?

Reply:
ah yes this comment should be generalized